edx/cc2olx
# [BD-27] [TNL-7394] Add Support for Assessments

1. Added support for parsing and converting of QTI v1.2 resources (`imsqti_xmlv1p2/imscc_xmlv1p1/assessment`  resource type).
2. Made some refactoring for `OlxExport`.
3. Added more documentation.
4. Updated tests and fixtures to cover new QTI-related code.

**JIRA tickets**: 
- [TNL-7394](https://openedx.atlassian.net/browse/TNL-7394)
- [BLENDED-539](https://openedx.atlassian.net/browse/BLENDED-539)

**Testing instructions**:

##### Validate imscc_file fixture
To be sure fixture is correct, we need to zip it's data and check if https://common-cartridge-viewer.netlify.app/ can parse and view it's content.

1.
    ```sh
    cd tests/fixtures_data/imscc_file
    ```
2. 
    ```sh
    zip -r test.imscc *
    ```
3. Upload `test.imscc` to https://common-cartridge-viewer.netlify.app/. You should see content of file with one Quiz.

##### Validate studio_course_xml fixture

You should be able to import this fixture in studio.

1.
    ```sh
    cd tests/fixtures_data/studio_course_xml
    ```
2.
    ```sh
    tar -czvf test.tar.gz course.xml
    ```
3. Import `test.tar.gz` in studio. You should see course with three units, one of them should be multiple checkbox problem. Correct answer should be "B".


##### Testing cc2olx itself

1. Install `cc2olx` (`python setup.py install` inside virtual environment).
1. Download [this file](https://github.com/edx/cc2olx/files/5049678/e54f1e98e43b4b648228b80f8aa9b129.zip).
1. Change it's extension to `.imscc`.
1. Try to convert it: `cc2olx -i e54f1e98e43b4b648228b80f8aa9b129.imscc`
1. Try to import resulting `tag.gz` archive in studio.
1. You should see units with imported multiple choice problems. Correct answers also sould be correctly imported.

**Author notes and concerns**:

1. For now, I don't know how to process html markup inside of descriptions of items in `imscc`.
1. This PR implements limited support for `cc.multiple_choice.v0p1` problem type. Detailed description can be found [in docstring](https://github.com/edx/cc2olx/blob/aad6bfeba3069e0a17b572206f7e0f422d4ae2a1/src/cc2olx/models.py#L773-L775), but in brief: score evaluating can be complicated (though in examples I see only simple case), so requires new XBlock to be implemented on platform side. However, since I've encountered only simple cases in various imscc files, when score can be `100` or `0`, I managed to implement this as multiple checkbox problem in platform.
1. This PR implements support only for `cc.multiple_choice.v0p1` because discovery had been done for QTI >v1.2, but in actual common cartridge files we see QTI<=1.2, which supports only multiple choice from initial list of problems.
1. Code for parsing `cc.multiple_choice.v0p1` and `cc.multiple_response.v0p1` is pretty similar, but score evaluation rules for latter always complicated. I didn't find right problem type in platform to match this, so I think new XBlock should be written to cover this.
    Example:
    ```xml
    <resprocessing>
      <outcomes>
        <decvar maxvalue="100" minvalue="0" varname="SCORE" vartype="Decimal"/>
      </outcomes>
      <respcondition continue="No">
        <conditionvar>
          <and>
            <not>
              <varequal respident="response1">7224</varequal>
            </not>
            <varequal respident="response1">7554</varequal>
            <varequal respident="response1">6954</varequal>
            <not>
              <varequal respident="response1">7317</varequal>
            </not>
            <varequal respident="response1">8698</varequal>
          </and>
        </conditionvar>
        <setvar action="Set" varname="SCORE">100</setvar>
      </respcondition>
    </resprocessing>
    ```

**Reviewers**
